### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by Daverlo**

### CODE REVIEW

File: src/main/java/com/semmle/jira/addon/LgtmServlet.java [hunk 1]
```diff
@@ -44,13 +44,30 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
     String[] pathParts = pathInfo.split("/");
     String configKey = pathParts[1];
 
+    Config config = Config.get(configKey);
+    List<Error> configErrors = config.validate();
+    for (Error error : configErrors) {
```
Comment (aibaars) [inline src/main/java/com/semmle/jira/addon/LgtmServlet.java]: This is a bit strange validate returns a list of errors, but we only tell the user about the first.

File: src/main/java/com/semmle/jira/addon/LgtmServlet.java [hunk 2]
```diff
@@ -248,6 +256,14 @@ private void sendJSON(HttpServletResponse resp, int code, Object value) throws I
     Util.JSON.writeValue(resp.getOutputStream(), value);
   }
 
+  class InvalidRequestException extends Exception {
```
Comment (aibaars) [inline src/main/java/com/semmle/jira/addon/LgtmServlet.java]: could be made static.

File: src/main/java/com/semmle/jira/addon/LgtmServlet.java [hunk 3]
```diff
@@ -139,15 +128,32 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
...
+    if (!Util.signatureIsValid(secret, bytes, req.getHeader("x-lgtm-signature"))) {
+      throw new AccessControlException("Forbidden");
+    }
+
+    Request request;
+    JsonNode jsonRequest;
+    try {
+      jsonRequest = Util.JSON.readTree(bytes);
+      request = new Request(jsonRequest);
+    } catch (IOException e) {
```
Comment (aibaars) [inline src/main/java/com/semmle/jira/addon/LgtmServlet.java]: I think we better catch `JsonParseException` only. Also I find the `e.getCause().getMessage()` a bit strange. Probably best to just include `e.getMessage()` as the message.

File: src/main/java/com/semmle/jira/addon/LgtmServlet.java [hunk 4]
```diff
@@ -228,7 +235,8 @@ void applyTransition(
     }
     if (!success) {
       // There was an applicable transition, however, it could not be executed.
-      // This is expected if a Validator Function rejects the transition. There is no point
+      // This is expected if a Validator Function rejects the transition. There is no
+      // point
```
Comment (aibaars) [inline src/main/java/com/semmle/jira/addon/LgtmServlet.java]: strange line break in comment.

File: src/test/java/com/semmle/jira/addon/TestValidateRequest.java [hunk 5]
```diff
@@ -1,35 +1,47 @@
...
-    Assert.assertEquals(
-        "{\"code\":403,\"error\":\"Forbidden.\"}", resp.getOutputStream().toString());
+public class TestValidateRequest extends TestLgtmServletBase {
+    HttpServletRequest req = mock(HttpServletRequest.class);
+	
+	@Before
+	public void setup() throws JsonGenerationException, JsonMappingException, IOException {
+		Request request = TestCreateIssue.createRequest("test", "Query", "test.cpp", "Security Error");
+		InputStream byteArrayInputStream = IOUtils.toInputStream(Util.JSON.writeValueAsString(request), "UTF-8");
+	    ServletInputStream servletInputStream=new ServletInputStream(){
```
Comment (aibaars) [inline src/test/java/com/semmle/jira/addon/TestValidateRequest.java]: Needs reformat.

File: src/main/java/com/semmle/jira/addon/LgtmServlet.java [hunk 6]
```diff
@@ -43,94 +45,119 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
...
+        case PROJECT_NOT_FOUND:
+        case USER_NOT_FOUND:
+          sendError(resp, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "Invalid configuration");
+          return;
+      }
     }
 
-    Request request;
-    JsonNode jsonRequest;
+    Request request = null;
```
Comment (aibaars) [inline src/main/java/com/semmle/jira/addon/LgtmServlet.java]: No need to set `request` to `null` right?

File: src/main/java/com/semmle/jira/addon/LgtmServlet.java [hunk 7]
```diff
@@ -43,94 +45,119 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
...
-        applyTransition(issue, Constants.WORKFLOW_SUPPRESS_TRANSITION_NAME, false, user, resp);
-        break;
-      case UNSUPPRESS:
-        applyTransition(issue, Constants.WORKFLOW_REOPEN_TRANSITION_NAME, true, user, resp);
-        break;
+      boolean skipValidation = request.transition != Transition.SUPPRESS;
+      String transitionName = null;
+      switch (request.transition) {
+        case CREATE:
+          // Already handled
```
Comment (aibaars) [inline src/main/java/com/semmle/jira/addon/LgtmServlet.java]: This comment is a bit funny. If we end-up in this case that means it actually wasn't handled. Perhaps better to throw an exception or just leave this case out, as the default case handles it quite well.

File: src/test/java/com/semmle/jira/addon/TestValidateRequest.java [hunk 8]
```diff
@@ -1,129 +1,48 @@
...
-    when(constantsManager.getStatus(any(String.class))).thenReturn(status);
-
-    // Config
-    setupConfig();
-    when(transactionTemplate.execute(ArgumentMatchers.<TransactionCallback<Config>>any()))
-        .thenReturn(config);
+  public void setup() throws JsonGenerationException, JsonMappingException, IOException {
+    Request request = TestCreateIssue.createRequest("test", "Query", "test.cpp", "Security Error");
+    InputStream byteArrayInputStream =
+        IOUtils.toInputStream(Util.JSON.writeValueAsString(request), "UTF-8");
```
Comment (aibaars) [inline src/test/java/com/semmle/jira/addon/TestValidateRequest.java]: `new ByteArrayInputStream(Util.JSON.writeValueAsBytes(request)`

File: src/main/java/com/semmle/jira/addon/config/Config.java [hunk 9]
```diff
@@ -1,125 +1,157 @@
...
-    return externalHookUrl;
+    if (properties.containsKey(PROPERTY_NAME_EXTERNAL_HOOK_URL))
+      return URI.create(properties.getProperty(PROPERTY_NAME_EXTERNAL_HOOK_URL));
+    return null;
   }
 
-  public void setExternalHookUrl(URI url) {
-    this.externalHookUrl = url;
+  @JsonProperty
+  public void setExternalHookUrl(String externalHookUrl) {
```
Comment (aibaars) [inline src/main/java/com/semmle/jira/addon/config/Config.java]: Please change back to `public void setExternalHookUrl(URI url)`

**Comment by Daverlo** [comment]
The last build failures are caused becase it can't download `https://m2proxy.atlassian.com/repository/public/org/apache/maven/maven-plugin-api/3.0/maven-plugin-api-3.0.pom`, but older builds have downloaded it successfully, and the url is accessible and downloads the correct POM file. What should we do?

**Comment by aibaars** [comment]
Thanks! Merging.

**Pull request merged**